docs/design/async-checkpoint.md

```diff
@@ -120,6 +124,93 @@ training process to resume the training, the training process can load the check
 from the shared memory not from the storage. Loading from the memory is much faster
 than the storage.
 
+## The ElasticAgent Asynchronously Saves the Checkpoint into Storage
+
+If we start a daemon subprocess of the GPU training process, the daemon
+subprocess will exit if the training process fails. In this case, the
+parameters in CPU memory of the daemon subprocess will be cleaned. So, the elastic agent
+in the main process can allocate the shared memory with the training
+process. The training process saves the state dict to the shared memory and
+the agent save them into the storage.
+
+### The Classes Design
+
+As we need a global thread to keep and sync the checkpointing state in storage, and an agent
+thread per node to save the checkpointing state into the storage, we design three classes to
+implement the checkpointing process.
+
+<div align="center">
+<img src="../figures/async-ckpt-classes.jpg" alt="Async Checkpoint Classes" width="1000">
+</div>
+
+- **GlobalCkptManager**
+    - Only one instance runs in the Rank 0 Agent process.
+    - The class is responsible for keeping the consistency of the checkpointing state in the shared
+      memory and the storage.
+    - Get the read lock of shared memory and notify all agents to save the checkpointing state
+      into the storage, update the checkpointing tracker file in the storage.
+- **AgentCkptManger**
+    - One instance runs in each agent process.
```

Comment:
It also need to allocate the shared memory blocks for each rank according to the size of checkpointing state dict.